### ISSUE TITLE
Multiple markers with/without a sprite, not showing correct one

### ISSUE THREAD

**Issue by tobrun** [mapbox/mapbox-gl-native#3239]
### Problem and output
When you add a marker for the second time to a map, it still shows the sprite from the previous one. 

When trying to do so I'm seeing this log message:
> W/mbgl: {Map}[Sprite]: Can't change sprite dimensions for 'com.mapbox.sprites.sprite_1'

### Use cases
I have tested and verified following workflows:

- First add a default marker than add a marker with a sprite
  - add a marker with no sprite shows the default marker
  - remove that marker
  - adding a marker with a sprite still shows default marker

- First add a marker with a sprite than add a default marker without a sprite
  - add a marker with a sprite shows the sprite
  - remove that marker  
  - adding a marker without a sprite still shows the sprite and not the default marker

### Code example
I received this code example form an end user
```java
public class MainActivity extends AppCompatActivity implements View.OnClickListener {

    private Marker marker;
    private MapView map;
    private Sprite sprite;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        map = (MapView) findViewById(R.id.map);
        map.setCenterCoordinate(new LatLng(37.76, -122.42));
        map.setStyleUrl(Style.LIGHT);
        map.setZoomLevel(11);
        map.onCreate(savedInstanceState);

        SpriteFactory spriteFactory = new SpriteFactory(map);
        Drawable icon = ContextCompat.getDrawable(getApplicationContext(), R.drawable.marker_green);
        sprite = spriteFactory.fromDrawable(icon);

        drawMarker(sprite);
        //drawMarker(null);

        findViewById(R.id.hideMarker).setOnClickListener(this);
        findViewById(R.id.showMarker).setOnClickListener(this);
    }

    public void onClick(View v) {
        switch (v.getId()) {
            case R.id.hideMarker:
                map.removeAllAnnotations();
                break;
            case R.id.showMarker:
                drawMarker(null);
                //drawMarker(sprite);
                break;
        }
    }

    private void drawMarker(Sprite sprite) {
        MarkerOptions mo = new MarkerOptions();
        mo.position(new LatLng(37.76, -122.42));

        if (sprite != null) {
            mo.icon(sprite);
        }

        if (map != null) {
            marker = map.addMarker(mo);
        }
    }
```

**Comment by tnightingale** [comment]
I suspect this bug is the result of MapView having a reference to its own SpriteFactory instance which is different to the one created in the code snippet above. The MapView has no knowledge of the new SpriteFactory and continues to refer to it's internal instance when rendering/managing markers.

A fix *might* be to update the MapView's local SpriteFactory reference?

Though I suspect that would cause issues if a new SpriteFactory is created late in the MapView's lifecycle, after it has had markers with custom icons added.

**Comment by tobrun** [comment]
merged with #3344
